src/bp/core/services/workspace-service.ts

```diff
@@ -49,156 +34,166 @@ export class WorkspaceService {
+      return (workspace && workspace.bots) || []
     }
   }
 
-  async assertUserExists(email: string): Promise<void> {
-    const user = await this.findUser({ email })
-    if (!user) {
-      throw new Error(`User "${email}" is not a part of the workspace`)
-    }
+  async findWorkspace(workspaceId: string): Promise<Workspace | undefined> {
+    const all = await this.getWorkspaces()
+    return all.find(x => x.id === workspaceId)
   }
 
-  async findUser(where: {}, selectFields?: Array<keyof AuthUser> | '*'): Promise<Partial<AuthUser> | undefined> {
-    const workspace = await this.getWorkspace()
-    const user = _.head(_.filter<AuthUser>(workspace.users, where))
-
-    if (!user) {
-      return undefined
-    }
-
-    if (selectFields === '*') {
-      return user
+  async createWorkspace(workspaceId: string, workspaceName: string): Promise<void> {
+    const workspaces = await this.getWorkspaces()
+    if (workspaces.find(x => x.id === workspaceId)) {
+      throw new Error(`Workspace with id "${workspaceId}" already exists`)
     }
 
-    return _.pick<AuthUser>(user, ...(selectFields || DEFAULT_USER_ATTRIBUTES)) as Partial<AuthUser>
+    workspaces.push({ ...DEFAULT_WORKSPACE, id: workspaceId, name: workspaceName })
+    return this.save(workspaces)
   }
 
-  async findRole(roleId: string): Promise<AuthRole> {
-    const workspace = await this.getWorkspace()
-    const role = workspace.roles.find(r => r.id === roleId)
-    if (!role) {
-      throw new Error(`Role "${roleId}" does not exists in workspace "${workspace.name}"`)
+  async addWorkspaceAdmin(email: string, strategy: string, workspaceId: string) {
+    const workspace = await this.findWorkspace(workspaceId)
+    if (workspace) {
+      return this.addUserToWorkspace(email, strategy, workspace.id, workspace.adminRole)
     }
-    return role
   }
 
-  async getRoleForUser(email: string): Promise<AuthRole | undefined> {
-    const user = await this.findUser({ email })!
-    return user && this.findRole(user.role!)
+  async addWorkspaceUser(email: string, strategy: string, workspaceId: string) {
+    const workspace = await this.findWorkspace(workspaceId)
+    if (workspace) {
+      return this.addUserToWorkspace(email, strategy, workspace.id, workspace.defaultRole)
+    }
   }
 
-  async createUser(authUser: BasicAuthUser | ExternalAuthUser): Promise<AuthUser> {
-    const workspace = await this.getWorkspace()
-    const newUser = {
-      ...authUser,
-      created_on: new Date()
-    } as AuthUser
-
-    // If there's no users, make the first account's role as Admin
-    if (!workspace.users.length) {
-      newUser.role = workspace.adminRole
-      await this.configProvider.mergeBotpressConfig({ superAdmins: [newUser.email] })
+  async addUserToWorkspace(email: string, strategy: string, workspace: string, role: string) {
+    const user = {
+      email,
+      strategy,
+      workspace,
+      role
     }
+    await this.workspaceRepo.createEntry(user)
```

Comment:
are potential errors handled ?

Reply:
They are handled at the upper level, since at that stage all is left is to insert in the database